src/Dapr.PluggableComponents/Components/PubSub/PubSubPublishRequest.cs

```diff
@@ -34,4 +36,14 @@ public sealed record PubSubPublishRequest(string PubSubName, string Topic)
     /// Gets the content type of the message.
     /// </summary>
     public string? ContentType { get; init; }
+
+    internal static PubSubPublishRequest FromPublishRequest(PublishRequest request)
+    {
+        return new PubSubPublishRequest(request.PubsubName, request.Topic)
+        {
+            ContentType = request.ContentType,
+            Data = request.Data.Memory,
+            Metadata = request.Metadata
```

Comment:
Right, the `init` on the property tells the compiler that it can only be set once, as part of construction, and at no other time.  You *can* still have constructor arguments that are optional (i.e. have defaults), though.  However, in that case, then the optional values are positional rather than named, which tends to get awkward when there are multiple optional values and where developers might use different subsets.